plugins/provisioners/salt/default_downloader.rb

```diff
@@ -0,0 +1,93 @@
+require 'open-uri'
+require 'digest'
+require_relative "./errors"
+
+module VagrantPlugins
+  module Salt
+    class DefaultDownloader
+      BASE_URL = "https://repo.saltproject.io/bootstrap/stable/"
+      PS1 = "bootstrap-salt.ps1"
+      SH  = "bootstrap-salt.sh"
+      SHA256_SUFFIX = "sha256"
+      SIG_SUFFIX = "asc"
+      KEY_URL = "#{BASE_URL}SALT-PROJECT-GPG-PUBKEY-2023.gpg"
+
+      def initialize(communicator)
+        @communicator = communicator
+        @logger  = Log4r::Logger.new("vagrant::salt::default_downloader")
+      end
+
+      def source_url 
+        @communicator == :winrm ? "#{BASE_URL}#{PS1}" : "#{BASE_URL}#{SH}"
+      end
+
+      def get_bootstrap_script
+        @logger.debug "Downloading bootstrap script from #{source_url}"
+        script_file = download_to_tmpfile(source_url)
+
+        if system("gpg --version")
+          @logger.debug "Verifying gpg signature"
+          verify_signature(script_file)
+        else
+          @logger.debug "Skipping gpg signature verification: gpg not found"
+        end
+
+        verify_sha256(script_file)
+
+        @logger.info "Downloaded and verified salt-bootstrap script "
+        script_file
+      end
+
+      def verify_sha256(script)
+        @logger.debug "Downloading sha256 file from #{source_url}/sha256"
+        sha256_file = download_to_tmpfile("#{source_url}.#{SHA256_SUFFIX}")
+        sha256 = extract_sha256(sha256_file.read)
+
+        @logger.debug "Computing sha256 value from script file"
+        computed_sha256 = Digest::SHA256.hexdigest(script.read)
+        script.rewind
+
+        @logger.debug "Comparing sha256 values"
+        if computed_sha256 != sha256
+          raise Salt::Errors::SaltError, :unverifiable_download
```

Comment:
Any other error handling tips for here?

Reply:
Creating a new error type (subclassing `SaltError`) would be good. Something like `InvalidShasumError` or something. The remote location of the script, expected sha value, and calculated sha value should also be included so the information can be displayed in the error message.